src/plan_expand_hypertable.c

```diff
@@ -846,23 +846,59 @@ collect_quals_walker(Node *node, CollectQualCtx *ctx)
 	return expression_tree_walker(node, collect_quals_walker, ctx);
 }
 
-static List *
-find_children_oids(HypertableRestrictInfo *hri, Hypertable *ht, LOCKMODE lockmode)
+static int
+chunk_cmp_chunk_id(const void *c1, const void *c2)
 {
-	/*
-	 * Using the HRI only makes sense if we are not using all the chunks,
-	 * otherwise using the cached inheritance hierarchy is faster.
-	 */
-	if (!ts_hypertable_restrict_info_has_restrictions(hri))
-		return find_inheritance_children(ht->main_table_relid, lockmode);
+	return (*(Chunk **) c1)->fd.id - (*(Chunk **) c2)->fd.id;
+}
+
+static Chunk **
+find_children_chunks(HypertableRestrictInfo *hri, Hypertable *ht, LOCKMODE lockmode,
+					 unsigned int *num_chunks)
+{
+	if (TS_HYPERTABLE_IS_INTERNAL_COMPRESSION_TABLE(ht))
+	{
+		/*
+		 * FIXME somehow chunk lookup doesn't work for internal compression
+		 * tables
+		 */
+		List *chunk_oids = find_inheritance_children(ht->main_table_relid, lockmode);
+		if (chunk_oids == NIL)
+		{
+			*num_chunks = 0;
+			return 0;
+		}
+
+		*num_chunks = list_length(chunk_oids);
+		Chunk **chunks = (Chunk **) palloc(sizeof(Chunk *) * *num_chunks);
 
+		for (int i = 0; i < *num_chunks; i++)
```

Comment:
nit: not sure if we are allowed to use C99 type definition inside `for` statement :)

Reply:
Let's use it, it looks better. I think even Postgres allows it now.